### ISSUE TITLE
forwardRef operator or it's analog proposal

### ISSUE THREAD

**Issue by thekip** [effector/effector#668]
Because of architecture of an Effector where all units and operations are written in top level of ES modules, and they start initialization in a moment when ES modules are imported you may often fall into trouble with circular dependencies and therefore broken build with "ReferenceError: Cannot access "$myStore" before initializaion" or even worst cases which sometimes very hard to debug and fix.  

## Proposal

The proposal is introducing a `forwardRef` concept: 

```typescript
sample({
 source: forwardRef(() => $myStore), // or simply define fn instead of references to other stores
 target: '...'
})
```

The forward ref fn is called somewhere on later phase when all ES modules tree is already initialized (batch and flush on next micro task, or synchronously execute when very first event is dispatched).

I saw many suggestions and solutions how to structure the code to avoid this kind of error, many of them far from ideal and require an additional boilerplate. Sometimes even when everything structured in a right way you still have circular deps problem because of complicated structure of components and stores, wich really hard to follow and find where exactly problem come.

**Comment by sergeysova** [comment]
I personally recommend to use [dependency-cruiser](https://github.com/sverweij/dependency-cruiser) to detect circular dependency.

This problem is a common issue for javascript projects, not only effector based. I think it is should be solved via architectural refactoring, not via some service code. Because, effector commonly is about business-logic

**Comment by zerobias** [comment]
How will you determine _when_ to execute your function? I can only see one reliable approach: execute them during the first fork call, which implies that it should be used. I'm not saying this is bad or good, I'm just wondering if it meets your expectations?

**Comment by thekip** [comment]
> This problem is a common issue for javascript projects, not only effector based. 

I totally not agree with this sentence. I've never had such problems in pure React projects nor Angular or any NodeJs server-side projects.
The problem lays in the main concept of Effector and how it binds one unit to another. Effector use side-effect approach of ES modules, which is basically in many ways considered as a bad practice. Because the logic depends on the fact just whether you import file or not (not necessary you import something from this file or not).

So the point of this issue it's, give users a way to somehow avoid problems related to this side-effect nature of Effector and make DX experience better avoiding playing with ES dependency tree or and moving imports back and forth.

**Comment by thekip** [comment]
> How will you determine when to execute your function? I can only see one reliable approach: execute them during the first fork call, which implies that it should be used. I'm not saying this is bad or good, I'm just wondering if it meets your expectations?

That's sounds good. If this would be called **after** ES deps tree fully initialized, that would solve the issue.

**Comment by zerobias** [comment]
Well, in that case, what happens during an explicit unit call? The situation is divided into two:
1) Pending functions will not be called. The unit will be launched in an inconsistent, partially initialized application
2) Pending functions will be called. Significant chance of getting the circular dependency back

**Comment by thekip** [comment]
What do you mean by `explicit unit call`? I'm not aware of Effector internals and I actually don't have a solution. I just want to rise this problem to community and start a discussion with a maintainers. 
In this issue, I want to say out loud that this problem exists and make DX with Effector sometimes horrible comparing to other solutions.

**Comment by zerobias** [comment]
Explicit direct call is just `event()` somewhere in the code

**Comment by zerobias** [comment]
Another edge case:

```ts
const result = sample({
  source: forwardRef(() => {
    return Math.random() > 0.5
      ? $storeA
      : eventB
  }),
  clock: $storeC
})
console.log(is.store(result))
```

What we will get in console log? Note that this case is not about determinism but about loosing some important properties of application structure

**Comment by thekip** [comment]
If you mean by this: 

```
const domain = createDomain();
const myEvent = domain.createEvent();

sample({
 clock: forwardRef(() => myEvent)
 target: '...'
})

myEvent() <-- Throw error, please call `domain.compile()` first or fork() or etc
```

**Comment by thekip** [comment]
> Another edge case:
> 
> ```ts
> const result = sample({
>   source: forwardRef(() => {
>     return Math.random() > 0.5
>       ? $storeA
>       : eventB
>   }),
>   clock: $storeC
> })
> console.log(is.store(result))
> ```
> 
> What we will get in console log? Note that this case is not about determinism but about loosing some important properties of application structure
> 
> In other words, how you will know what forwardRef will return ahead of time?

This should be not supported. Typings should be written in the way were a strongly typed Tuple is returned. ESlint rule can  check whether a conditional or any other logic presence in forwardRef and ban it.

**Comment by zerobias** [comment]
I mean, _how_ sample should know that we returning from forwardRef? If we return store then result will be store to, but how you will explain that to code still executed at the root?

**Comment by zerobias** [comment]
There are a lot of methods in the library and it’s ecosystem which choose what to do with its arguments based on their types.
In short, your solution should return something meaningful in that case:

```ts
console.log(is.store(forwardRef(() => $foo)))
```

And note that it still should be known immediately

**Comment by thekip** [comment]
I have no idea, actually I think sample should not be aware of that and should treat all passed units as polymorphic entities and do the same stuff for all. 

```typescript
console.log(is.store(forwardRef(() => $foo)))
```

Or if this  used somewhere in internals, all internals should be changed to be executed only after compile() / fork is called. Means we delay / defer all chain building until the moment of calling explicitly, not as a side effect of ES module. 

This is not going to be a simple change indeed, but the result would have great impact on overall DX.

**Comment by zerobias** [comment]
`is.store` returns its result immediately, that should be taken into account. I understand and support your goals, but this just show that current proposal has significant unsolved issue. Look at how es proposals works: champion of the proposal talk with es committee, and if they together find some edge cases, he improves that and comes back

**Comment by thekip** [comment]
What are the use cases of `is.store` outside of libraries internals?

**Comment by zerobias** [comment]
Every other case could be reduced to this, important one


You may have already noticed that laziness percolate the whole concept and brings quite a few unique issues, so this is not an easy proposal to suggest, otherwise it would be solved a long time ago, so you’ll need your unconventional thinking

**Comment by thekip** [comment]
As i mentioned, internals of library such as sample, forward, etc might be just rewritten to support "delayed" pattern. Such as 

```typescript
function sample(...args) {
  getDomain(args).batch.push(() => {
  // use is.* methods here 
  // smth like
  if (args.source().some(is.store)) {
   // do whatever with store
  } 
})

return new LazyUnit(...)
}
```
And if user whants to use `is.*` methods in it's code, so it's up to his scope of responsibility to use it after iniitalizations. 

So: 

```typescript

const $derivedStore = sample({
  source: forwardRef(() => [$store1, store2])
}) // returns LazyUnit

is.store($derivedStore) // <-- error, you trying to get type of Lazy store before it's compiled

// this a static store, we can resolve immedieately with no error,
// if user call it in the phase of ES modules initializations it's users responsibility to make it work. 
is.store($store1)

domain.compile()

is.store($derivedStore) // <-- success, returned `true`
```

**Comment by thekip** [comment]
Just came up with another idea: 

```typescript
const $store3: Deferred<Store> = domain.defer(($_store, $_store2) => sample({
    source: [$_store],
    target: [$_store2],
  }),
  () => [$store, $store2]
);

const $store4 = domain.defer(($_store3) => sample({
    source: [$_store3],
    ...
  }),
  () => [$store3]
);


// somewhere in app, where fork is used

domain.compile() // execute all defer() functions and bind units to each other
```

This a `defer()` api. This concept could be implemented as completely standalone api without changing internals at all. 

```typescript
// pseudo typescript typing
defer<T extends Tuple<Unit>>(factory: (...T[]) => Unit, deps: () => (T | Deferred<T>)[]);
```

We just wrap every call where we depend on other units in `defer()` and every call where result of `defer()` call is used.

`defer()` itself returns  `Deferred<Unit>` which by the fact just a function which will resolve all dependencies and pass it into a `factory()`.

We also need to change bindings such a useStore() to be able to 'unwrap' deferred values. 

1. The **benefit** of this api - it might be completely separated from Effector. And even implemented as separate package. 
2. The **drawback** is, it's quite noisy, because you need to write your dependencies twice and thinking of the names for variables inside a factory. But this might be easily solved by babel plugin, simplifying api to just:

```typescript
const $store3: Deferred<Store> = domain.defer(() => sample({
    source: [$store],
    target: [$store2],
  }),
);
```

**Comment by zerobias** [comment]
Everything could be easier if we allow lazy initialization only for operators which don’t create new units, assume to have a `target` field. Then yes, we might use separate `defer` api or built in `forwardRef`

**Comment by zerobias** [comment]
> This concept could be implemented as completely standalone api

I SUDDENLY  realized that we have such api, it called "Promise" 🙃

So, take a part about `target` and a part about promises and we have a way to resolve an issue: effector should support promises in operators with `target`, somehow like that:

```ts
sample({
  source: import('./user').then(({$name}) => $name),
  target: import('./analytics').then(({sendFx}) => sendFx),
})
```

**Comment by thekip** [comment]
What about `is.store` which should return immediately? If this requirement is dropped, so it's better to implement one of the previous proposals (with forward ref or function) because promise based solution looks like a fixing problem with a wrong instrument.

> if we allow lazy initialization only for operators which don’t create new units

If we assume that, then `defer` function would be much easier. 

```ts
// don't need to resolve dependecies, don't need to introduce any new entities, etc
domain.defer(() => sample({
    source: [$store],
    target: [$store2],
  }),
);
```

**Comment by zerobias** [comment]
> What about is.store which should return immediately?

If you want to create new unit you will not be able to use neither promises nor `domain.defer` anyway

So your code will looks like that

```ts
import {delay} from 'patronum'
import {domain} from './root'
import {triggerA, triggerB, requestFx} from './plain'
import {$fxDelay, delayedTriggerB} from './cyclic'

// this unit have everything to be created at the root
export const delayedTriggerA = delay({
  source: triggerA,
  timeout: 50
})

// this unit need cyclic $fxDelay
export const delayedRequest = createEvent()

domain.defer(() => {

  // link for unit in that module
  delay({
    source: delayedRequest,
    timeout: $fxDelay,
    target: requestFx,
  })

  // link for unit in other module
  delay({
    source: delayedTriggerB,
    timeout: 50,
    target: triggerB,
  })

})
```

**Comment by thekip** [comment]
So do we agree on this proposal?

**Comment by zerobias** [comment]
Yes. But I'm not sure about name and relation with domain. Maybe domain is not required?

```ts
import {delay} from 'patronum'
import {defer} from 'effector'
import {triggerA, triggerB, requestFx} from './plain'
import {$fxDelay, delayedTriggerB} from './cyclic'

// this unit have everything to be created at the root
export const delayedTriggerA = delay({
  source: triggerA,
  timeout: 50
})

// this unit need cyclic $fxDelay
export const delayedRequest = createEvent()

defer(() => {

  // link for unit in that module
  delay({
    source: delayedRequest,
    timeout: $fxDelay,
    target: requestFx,
  })

  // link for unit in other module
  delay({
    source: delayedTriggerB,
    timeout: 50,
    target: triggerB,
  })

})
```

**Comment by thekip** [comment]
I've just followed pattern with other functions such as `domain.createEvent`, `domain.createEffect` etc. I'm a new user of effector an if this is not required and will work with no issue in SSR and tests so why not.

**Comment by zerobias** [comment]
Okay, then now we need to do a smoke test: is it enough to move to `defer` only links with `target`?

Let's assume naive defer implementation (it will not be async in prod btw):

```ts
export function defer(fn) {
  return Promise.resolve().then(fn)
}
```

**Please try to use it in your projects**. Is it useful, did you avoided circular loops? Have you been able to use it in all the places where you would like to? Note that methods that create new units will have to be rewritten to an explicit unit creation and method with `target`
